Support strictly blittable type marshalling (#69100)
dotnet/runtime: docs/design/libraries/LibraryImportGenerator/StructMarshalling.md

For the auto-layout clause, we have one small issue; today, our ref-assemblies do not expose if a value type is marked as `[StructLayout(LayoutKind.Auto)]`, so we'd still have some cases where we might have runtime failures. However, we can update the tooling used in dotnet/runtime, GenAPI, to expose this information if we so desire. Once that case is handled, we have a mechanism that we can safely use to determine, at compile time, which types will not require marshalling. If we decide to not cover this case (as cases where users mark types as `LayoutKind.Auto` manually are exceptionally rare), we still have a solid design as Roslyn will automatically determine for us if a type is `unmanaged`, so we don't need to do any additional work.

Users are permitted to rely upon the built-in notion of blittable if the type to marshal adheres to the following constraints, termed "strictly blittable" in code:

1) Is always blittable (for example, `int` or `double`, but not `char`).
2) Is a value type defined in the source project the current source generator is running on.
3) Is a type composed of types adhering to (1) and (2).

> The "strictly blittable" notion is being introduced to ease adoption. The impact of `DisableRuntimeMarshallingAttribute` is that some scenarios require effort that is much larger relative to the need to marshal an internal transfer type that has historically been blittable. For example, the following value type would require the consumer of the source generator to apply `DisableRuntimeMarshallingAttribute` if the above was not permitted.

Reviewer: ```suggestion
The "strictly blittable" notion is being introduced to ease adoption. The impact of `DisableRuntimeMarshallingAttribute` can be quite large, impeding adoption in scenarios where the need is only to marshal an internal transfer type that has historically been blittable. For example, the following value type would require the consumer of the source generator to apply `DisableRuntimeMarshallingAttribute` if the above was not permitted.
```

> The "strictly blittable" notion is being introduced to ease adoption. The impact of `DisableRuntimeMarshallingAttribute` is that some scenarios require effort that is much larger relative to the need to marshal an internal transfer type that has historically been blittable. For example, the following value type would require the consumer of the source generator to apply `DisableRuntimeMarshallingAttribute` if the above was not permitted.

Reviewer 2: > some scenarios require effort that is much larger relative to the need to marshal an internal transfer type that has historically been blittable.

This contradicts statement made in the next paragraph "As we believe that users should be able to move over their assemblies to the new source-generated interop world as a whole assembly".

Author: Great point all around. Will update.
